### PR TITLE
[DependencyInjection] Fix autowiring collisions detection

### PR DESCRIPTION
**Pull request by GuilhemN**
| Q             | A
| ------------- | ---
| Branch?       | 2.8
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        |

Fixes https://github.com/symfony/symfony/pull/21658 by implementing the second proposal of https://github.com/symfony/symfony/pull/21658#issuecomment-280858452:
> Another idea: store the types used previously and check that new services registered don't implement them.

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
The benefit here is simplicity, the drawback is that this won't work with setter autowiring, where instead of throwing, we skip. Can you try something on master also see if you can make this approach fit there or if we need v1?

**Comment by GuilhemN** [comment]
@nicolas-grekas you're talking about the fact that an exception could be throw when using a setter, right? Imo this is a very weird case and I don't think we should take position here, I think it's better to throw an exception and let the user manually fix the issue. For example, should a service autowired in a constructor win against a setter?

**Comment by nicolas-grekas** [comment]
My bad: on conflicts, an exception is always thrown even for setters, so should be good.
But the check should be done outside of the autowiring loop - any ambiguous type must throw - setter or not.

File: src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php [hunk 1]
```diff
@@ -44,6 +45,15 @@ public function process(ContainerBuilder $container)
                     $this->completeDefinition($id, $definition);
                 }
             }
+
+            foreach ($this->usedTypes as $type => $id) {
+                if (isset($this->usedTypes[$type]) && isset($this->notGuessableTypes[$type])) {
+                    $classOrInterface = class_exists($type) ? 'class' : 'interface';
+                    $matchingServices = implode(', ', $this->types[$type]);
+
+                    throw new RuntimeException(sprintf('Unable to autowire argument of type "%s" for the service "%s". Multiple services exist for this %s (%s).', $type, $id, $classOrInterface, $matchingServices));
```
Comment (GuilhemN) [inline src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php]: Only the last service in which the type is used will be shown, would you prefer a list of all of them?

Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php]: the last is enough to me

File: src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php [hunk 2]
```diff
@@ -44,6 +45,15 @@ public function process(ContainerBuilder $container)
                     $this->completeDefinition($id, $definition);
                 }
             }
+
+            foreach ($this->usedTypes as $type => $id) {
+                if (isset($this->usedTypes[$type]) && isset($this->notGuessableTypes[$type])) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php]: `&& !isset($this->definedTypes[$type])`?

**Review by nicolas-grekas — APPROVED** [review]
:+1:

File: src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php (hunk 1 above)
Comment (GuilhemN) [inline src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php]: Ok, let's keep it as is :)

File: src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php (hunk 2 above)
Comment (GuilhemN) [inline src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php]: That was a good question actually, that made me catch a bug (see https://github.com/symfony/symfony/pull/21670).

**Comment by fabpot** [comment]
Thank you @GuilhemN.

**Comment by GuilhemN** [comment]
Tell me if you need help to merge this in higher branches, it needs to be adapted (`notGuessableTypes` -> `ambiguousServiceTypes`, add getters autowiring support).

**Comment by fabpot** [comment]
@GuilhemN Indeed, if you can submit a PR for higher branches, that would help. Thanks.

**Pull request merged**